Marlin/stepper.cpp

```diff
@@ -796,6 +819,104 @@ void Stepper::isr() {
   #endif
 }
 
+void Stepper::chunk_steps() {
+  const uint8_t *cur_chunk = chunk_buffer[current_block->chunk_idx];
+  const uint8_t cur_chunk_block = uint8_t(step_events_completed >> 3) << 1;
+  const uint8_t block_steps = uint8_t(step_events_completed & 0x7);
+
+  const uint8_t a = cur_chunk[cur_chunk_block];
+  const uint8_t b = cur_chunk[cur_chunk_block + 1];
+
+  const uint8_t dX = a >> 4;
+  const uint8_t dY = a & 0xF;
+  const uint8_t dZ = b >> 4;
+  const uint8_t dE = b & 0xF;
+
+  uint8_t steps[4] = { 0 };
+  steps[X_AXIS] = block_moves[dX][(block_steps + 0) & 0x7];
+  steps[Y_AXIS] = block_moves[dY][(block_steps + 2) & 0x7];
+  steps[Z_AXIS] = block_moves[dZ][(block_steps + 4) & 0x7];
+  steps[E_AXIS] = block_moves[dE][(block_steps + 6) & 0x7];
+
+  //start of block, check direction
+  if(block_steps == 0) {
+    unsigned char dm = last_direction_bits;
+
+    #define UPDATE_DIR(AXIS) \
+	  if(d## AXIS == 0) {} \
+      else if(d## AXIS < 7) SBI(dm, AXIS ##_AXIS); \
+      else CBI(dm, AXIS ##_AXIS);
+
+    UPDATE_DIR(X);
+    UPDATE_DIR(Y);
+    UPDATE_DIR(Z);
+    UPDATE_DIR(E);
+
+    if (dm != last_direction_bits) {
+      last_direction_bits = dm;
+      set_directions();
+    }
+  }
+
+  #define PULSE_STARTC(AXIS) \
+    if (steps[_AXIS(AXIS)]) { _APPLY_STEP(AXIS)(!_INVERT_STEP_PIN(AXIS), 0); }
+
+  // Stop an active pulse, reset the Bresenham counter, update the position
+  #define PULSE_STOPC(AXIS) \
+    if (steps[_AXIS(AXIS)]) { \
+      count_position[_AXIS(AXIS)] += count_direction[_AXIS(AXIS)]; \
```

Comment:
in theory this should let us resume marlin based planning after doing chunks, provided the planner syncs the "real" position back to it.